uaclient/tests/test_config.py

```diff
@@ -0,0 +1,172 @@
+
+        assert False is os.path.exists(expected_path), (
+            'Found unexpected file %s' % expected_path)
+        assert None is cfg.write_cache(key, content)
+        assert True is os.path.exists(expected_path), (
+            'Missing expected file %s' % expected_path)
+        assert content == cfg.read_cache(key)
+
+    def test_write_cache_creates_dir_when_data_dir_does_not_exist(self):
+        """When data_dir doesn't exist, create it."""
+        tmp_subdir = self.tmp_path('does/not/exist')
+        cfg = UAConfig({'data_dir': tmp_subdir})
+
+        assert False is os.path.isdir(tmp_subdir), (
+            'Found unexpected directory %s' % tmp_subdir)
+        assert None is cfg.write_cache('somekey', 'someval')
+        assert True is os.path.isdir(tmp_subdir), (
+            'Missing expected directory %s' % tmp_subdir)
+        assert 'someval' == cfg.read_cache('somekey')
+
+    @params(('dictkey', {'1': 'v1'}), ('listkey', [1, 2, 3]))
+    def test_write_cache_writes_json_string_when_content_not_a_string(
+            self, key, value):
+        """When content is not a string, write a json string."""
+        tmp_dir = self.tmp_dir()
+        cfg = UAConfig({'data_dir': tmp_dir})
+
+        expected_json_content = json.dumps(value)
+        assert None is cfg.write_cache(key, value)
+        with open(self.tmp_path(key, tmp_dir), 'r') as stream:
+            assert expected_json_content == stream.read()
+        assert value == cfg.read_cache(key)
+
+
+class TestReadCache(TestCase):
+
+    @params(*KNOWN_DATA_PATHS)
+    def test_read_cache_returns_none_when_data_path_absent(
+            self, key, path_basename):
+        """The data_path method returns the data_dir when key is absent."""
+        cfg = UAConfig({'data_dir': '/my/dir'})
+        assert None is cfg.read_cache(key)
+        assert False is os.path.exists(os.path.join('/my/dir', path_basename))
+
+    @params(*KNOWN_DATA_PATHS)
+    def test_read_cache_returns_content_when_data_path_present(
+            self, key, path_basename):
+        tmp_dir = self.tmp_dir()
+        cfg = UAConfig({'data_dir': tmp_dir})
+        data_path = self.tmp_path(path_basename, tmp_dir)
+        with open(data_path, 'w') as f:
+            f.write('content%s' % key)
+
+        assert 'content%s' % key == cfg.read_cache(key)
+
+    @params(*KNOWN_DATA_PATHS)
+    def test_read_cache_returns_stuctured_content_when_json_data_path_present(
+            self, key, path_basename):
+        tmp_dir = self.tmp_dir()
+        cfg = UAConfig({'data_dir': tmp_dir})
+        data_path = self.tmp_path(path_basename, tmp_dir)
+        expected = {key: 'content%s' % key}
+        with open(data_path, 'w') as f:
+            f.write(json.dumps(expected))
+
+        assert expected == cfg.read_cache(key)
+
+
+class TestDeleteCache(TestCase):
+
+    @params('_contracts', '_entitlements', '_machine_token')
+    def test_delete_cache_unsets_any_local_cached_class_attributes(self, attr):
```

Comment:
In my opinion, this test is tied more closely to the implementation than is really necessary. I think we'd be better testing the invariants of the delete_cache operation (i.e. if things are set before, they are unset afterwards, and corresponding cache files are removed), so that if we decide to add an in-memory cache for another attribute (or, for an example that would require more than extending @params on this test, if we decide to move the in-memory caching to be in a single dict rather than separate attributes), we don't have to modify these tests.

Reply:
Changed up the tests a bit to validate cached propert cfg.entitlements and cfg.contracts public properties with previous settings get cleared by delete_cache